infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java

```diff
@@ -247,6 +261,56 @@ private void execute(
     }
   }
 
+  /**
+   * Watches failure events for the specified Pod, and deletes the Pod if a failure event is
+   * detected
+   *
+   * @param deployments the KubernetesDeployments object
+   * @param podName the name of the Pod to check events for
+   * @throws InfrastructureException
+   */
+  private void watchFailureEventAndTerminatePod(KubernetesDeployments deployments, String podName)
+      throws InfrastructureException {
+    deployments.watchEvents(
```

Comment:
I'm not sure who is the author of watchEvents, maybe even me... But AFAIU this event watcher instance is going to live forever, which would lead to more memory consumption and making server slower to process. Please consider contribution unsubscribe method to deployments where you pass watcher instance... Or make deployments return subscriptions where you can invoke unsubscribe in finally statement.

Reply:
The watch is being cleaned up here in the finally block:
https://github.com/eclipse-che/che-server/blob/90968896473bf0172d9dc143906a91b9f89c96aa/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/pvc/PVCSubPathHelper.java#L253-L255

I think this is ok?

Reply:
Huh, glad that it's in place!
I must to finish my cup of coffee before reviewing PR :-D